server/sonar-server/src/main/java/org/sonar/server/scanner/ProjectAction.java

```diff
@@ -0,0 +1,154 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.scanner;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.sonar.api.server.ws.Request;
+import org.sonar.api.server.ws.Response;
+import org.sonar.api.server.ws.WebService;
+import org.sonar.batch.protocol.input.FileData;
+import org.sonar.batch.protocol.input.ProjectRepositories;
+import org.sonarqube.ws.WsScanner.WsProjectResponse;
+
+import static org.sonar.server.ws.WsUtils.writeProtobuf;
+
+public class ProjectAction implements ScannerWsAction {
+
+  private static final String PARAM_KEY = "key";
+  private static final String PARAM_PROFILE = "profile";
+  private static final String PARAM_PREVIEW = "preview";
+
+  private final ProjectDataLoader projectDataLoader;
+
+  public ProjectAction(ProjectDataLoader projectDataLoader) {
+    this.projectDataLoader = projectDataLoader;
+  }
+
+  @Override
+  public void define(WebService.NewController controller) {
+    WebService.NewAction action = controller.createAction("project")
+      .setDescription("Return project repository")
+      .setSince("4.5")
+      .setInternal(true)
+      .setHandler(this);
+
+    action
+      .createParam(PARAM_KEY)
+      .setRequired(true)
+      .setDescription("Project or module key")
+      .setExampleValue("org.codehaus.sonar:sonar");
+
+    action
+      .createParam(PARAM_PROFILE)
+      .setDescription("Profile name")
+      .setExampleValue("SonarQube Way");
+
+    action
+      .createParam(PARAM_PREVIEW)
+      .setDescription("Preview mode or not")
+      .setDefaultValue(false)
+      .setBooleanPossibleValues();
+  }
+
+  @Override
+  public void handle(Request wsRequest, Response wsResponse) throws Exception {
+    ProjectRepositories data = projectDataLoader.load(ProjectDataQuery.create()
+      .setModuleKey(wsRequest.mandatoryParam(PARAM_KEY))
+      .setProfileName(wsRequest.param(PARAM_PROFILE))
+      .setIssuesMode(wsRequest.mandatoryParamAsBoolean(PARAM_PREVIEW)));
+
+    WsProjectResponse projectResponse = buildResponse(data);
+    writeProtobuf(projectResponse, wsRequest, wsResponse);
+  }
+
+  private static WsProjectResponse buildResponse(ProjectRepositories data) {
+    WsProjectResponse.Builder response = WsProjectResponse.newBuilder();
+    setLastAnalysisDate(response, data);
+    response.setTimestamp(data.timestamp());
+    response.getMutableFileDataByModuleAndPatch()
+      .putAll(buildFileDataByModuleAndPatch(data));
+    response.getMutableSettingsByModule()
+      .putAll(buildSettingsByModule(data));
+
+    return response.build();
+  }
+
+  private static void setLastAnalysisDate(WsProjectResponse.Builder response, ProjectRepositories data) {
+    if (data.lastAnalysisDate() != null) {
+      response.setLastAnalysisDate(data.lastAnalysisDate().getTime());
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) NullPointerException might be thrown as 'lastAnalysisDate' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2259)